src/app/GitUI/UserControls/RevisionGrid/FormQuickItemSelector.cs

```diff
@@ -1,101 +1,102 @@
+                longestItemWidth = Math.Max(longestItemWidth, graphics.MeasureString(label, lbxRefs.Font).Width);
             }
         }
-
-        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
+        finally
         {
-            if (keyData == Keys.Escape)
-            {
-                DialogResult = DialogResult.Cancel;
-                Close();
-                return true;
-            }
+            // what is this magic number "MaxVisibleItemsWithoutScroll + 0.5"?
+            // we are limiting number of visible items in the listbox before enabling vscroll, this is to reduce the risk of not fitting on user's screen.
+            // when listbox.IntegralHeight=true, the listbox automatically calculates its size and only show full items (i.e. you can't render it
+            // at 20px high if the ItemHeight=13px, it can only be 13, 26, 39 etc (NB: slightly more if you account for the furniture)).
+            // listbox.PreferredHeight returns the height of the listbox showing all items.
+            // for some strange reason, MaxVisibleItemsWithoutScroll*listbox.ItemHeight renders the listbox showing MaxVisibleItemsWithoutScroll-1 items.
+            // to fix this, trick the listbox and set its height to show MaxVisibleItemsWithoutScroll+0.5
+            // this internally forces it to re-calculates its bounds and then to calculate its preferred height
+            lbxRefs.Height = Math.Min(lbxRefs.PreferredHeight, (int)((MaxVisibleItemsWithoutScroll + 0.5) * lbxRefs.ItemHeight));
+            lbxRefs.Width = (int)longestItemWidth;
+            lbxRefs.EndUpdate();
+            lbxRefs.SelectedIndex = 0;
+            lbxRefs.Focus();
 
-            return base.ProcessCmdKey(ref msg, keyData);
+            // resize the form to look nice
+            Size = new Size(lbxRefs.Width + Padding.Left + Padding.Right,
+                            lbxRefs.Height + Padding.Top + flowLayoutPanel1.Height);
         }
+    }
 
-        private void lbxRefs_MouseDoubleClick(object sender, MouseEventArgs e)
+    protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
+    {
+        if (keyData == Keys.Escape)
         {
-            AcceptButton.PerformClick();
+            DialogResult = DialogResult.Cancel;
+            Close();
+            return true;
         }
 
-        public class ItemData
-        {
-            public string Label { get; }
-            public object Item { get; }
+        return base.ProcessCmdKey(ref msg, keyData);
+    }
 
-            public ItemData(string label, object item)
-            {
-                Label = label;
-                Item = item;
-            }
+    private void lbxRefs_MouseDoubleClick(object sender, MouseEventArgs e)
+    {
+        AcceptButton.PerformClick();
+    }
+
+    internal sealed class ItemData
+    {
+        public string Label { get; }
+        public object Item { get; }
+        public object Tag { get; }
```

Comment:
1. "Tag" is misleading.
```suggestion
        public string Category { get; }
```

2. There is absolutely no need to store the section name (`Category`) for every ref. The regarding text can be chosen when the section header is inserted. This could be done using a functor passed instead of all those `TranslationString`s.
 
3. (There is no real need for `ItemData` any longer. `ItemRef.Name` could be used in place.)

Reply:
Thanks.

1\. - yes, agree
2\. - yes, agree
3\. - can't agree on this one, at least it's not obvious to me.